src/DotLiquid/Context.cs

```diff
@@ -430,14 +430,15 @@ private object FindVariable(string key)
             if (scope == null)
             {
                 foreach (Hash e in Environments)
-                    if ((variable = LookupAndEvaluate(e, key)) != null)
+                    if (TryEvaluateHashOrArrayLikeObject(e, key, out variable))
                     {
                         scope = e;
                         break;
                     }
             }
             scope = scope ?? Environments.LastOrDefault() ?? Scopes.Last();
-            variable = variable ?? LookupAndEvaluate(scope, key);
+            if (variable == null)

```

Comment:
By not saving result of Try, a explicit null value in a dictionary will recompute. This is inefficient. See my PR how I rewrote the logic here, but feel free to offer another solution.

Reply:
Incorporated changes from @microalps dotliquid/render-perf branch.